packages/remix-server-runtime/serialize.ts

```diff
@@ -1,21 +1,60 @@
-import type { Jsonify } from "./jsonify";
+import type { EmptyObject, Jsonify } from "./jsonify";
 import type { TypedDeferredData, TypedResponse } from "./responses";
+import type {
+  ClientActionFunctionArgs,
+  ClientLoaderFunctionArgs,
+} from "./routeModules";
 import { expectType } from "./typecheck";
 import { type Expect, type Equal } from "./typecheck";
 
 // prettier-ignore
 /**
- * Infer JSON serialized data type returned by a loader or action.
+ * Infer JSON serialized data type returned by a loader or action, while
+ * avoiding deserialization if the input type if it's a clientLoader or
+ * clientAction that returns a non-Response
  *
  * For example:
  * `type LoaderData = SerializeFrom<typeof loader>`
  */
 export type SerializeFrom<T> =
-  T extends (...args: any[]) => infer Output ? Serialize<Awaited<Output>> :
+  T extends (...args: any[]) => infer Output ?
+    Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs] ?
+      // Client data functions may not serialize
+      SerializeClient<Awaited<Output>>
+    :
+    // Serialize responses
+    Serialize<Awaited<Output>>
+  :
   // Back compat: manually defined data type, not inferred from loader nor action
   Jsonify<Awaited<T>>
 ;
 
+// note: cannot be inlined as logic requires union distribution
+// prettier-ignore
+type SerializeClient<Output> =
+  Output extends TypedDeferredData<infer U> ?
+    // top-level promises
+    & {
+        [K in keyof U as K extends symbol
+          ? never
+          : Promise<any> extends U[K]
+          ? K
+          : never]: DeferValueClient<U[K]>; // use generic to distribute over union
+    }
+    // non-promises
+    & {
+      [K in keyof U as Promise<any> extends U[K] ? never : K]: U[K];
+    }
+  :
+  Output extends TypedResponse<infer U> ? Jsonify<U> :
+  Awaited<Output>
+
+// prettier-ignore
+type DeferValueClient<T> =
+  T extends undefined ? undefined :
+  T extends Promise<unknown> ? Promise<Awaited<T>> :
+  T;
+
```

Comment:
Copies of the `Serialize`/`DeferValue` types with `Jsonify` stuff removed (unless it's a `TypedResponse` and then we still call `Jsonify`)